bqskit/passes/util/structure.py

```diff
@@ -0,0 +1,41 @@
+"""This module implements the StructureAnalysisPass."""
+from __future__ import annotations
+
+from bqskit.compiler.basepass import BasePass
+from bqskit.compiler.passdata import PassData
+from bqskit.ir.circuit import Circuit
+from bqskit.ir.circuit import CircuitGate
+from bqskit.ir.structure import CircuitStructure
+
+
+class StructureAnalysisPass(BasePass):
+    """
+    A pass that catalogs circuit structures used in a partitioned circuit.
+
+    Note: For each partitioned CircuitGate in the Circuit, the recursive
+    `unfold_all` method is called to ensure that structure is defined at
+    the gate level. For structures to be considered the same, they must
+    consist of the same gates, in the same locations.
+    """
+
+    def __init__(self) -> None:
+        """Construct a StructurePass."""
+
+    async def run(self, circuit: Circuit, data: PassData) -> None:
+        """Perform the pass's operation, see :class:`BasePass` for more."""
+
+        if 'structures' not in data:
```

Comment:
What if `structures` is in data? Should that be a warning? or be overridden?

Reply:
Added warning. It will be overridden.